prov/mrail/src/mrail_init.c

```diff
@@ -131,27 +121,131 @@ int mrail_get_core_info(uint32_t version, const char *node, const char *service,
 
 	if (!core_hints->fabric_attr) {
 		core_hints->fabric_attr = calloc(1, sizeof(*core_hints->fabric_attr));
-		if (!core_hints->fabric_attr) {
-			ret = -FI_ENOMEM;
-			goto out;
-		}
+		if (!core_hints->fabric_attr)
+			goto err;
 	}
 
 	if (!core_hints->domain_attr) {
 		core_hints->domain_attr = calloc(1, sizeof(*core_hints->domain_attr));
-		if (!core_hints->domain_attr) {
-			ret = -FI_ENOMEM;
-			goto out;
-		}
+		if (!core_hints->domain_attr)
+			goto err;
 	}
 	core_hints->domain_attr->av_type = FI_AV_TABLE;
+	return core_hints;
+
+err:
+	fi_freeinfo(core_hints);
+	return NULL;
+}
+
+/*
+ * Gather entries from an array of per-rail info lists to form an array of multi-rail info.
+ * Each multi-rail info is a list consists of one info per rail.
+ */
+static int mrail_gather_rail_info(struct fi_info **rail_info, int num_rails,
+				  struct fi_info ***info_out, int *num_info_out)
+{
+	struct fi_info **mrail_info, **tails, *p;
+	int num_mrail_info = 0;
+	int i, j, n;
+
+	/*
+	 * find the shortest length of the per-rail list, this is the number of
+	 * multi-rail info available
+	 */
+	for (i = 0; i < num_rails; i++) {
+		n = 0;
+		p = rail_info[i];
+		while (p) {
+			n++;
+			p = p->next;
+		}
+		if (i == 0 || num_mrail_info > n)
+			num_mrail_info = n;
+	}
+
+	if (!num_mrail_info) {
+		FI_WARN(&mrail_prov, FI_LOG_FABRIC,
+			"Some rails cannot be used!\n");
+		return -FI_ENODATA;
+	}
+
+	mrail_info = calloc(num_mrail_info, sizeof(*mrail_info));
+	if (!mrail_info)
+		return -FI_ENOMEM;
+
+	tails = calloc(num_mrail_info, sizeof(*tails));
+	if (!tails) {
+		free(mrail_info);
+		return -FI_ENOMEM;
+	}
+
+	/*
+	 * Do the gathering -- Take one item from each rail and form a list.
+	 * Repeat until some rails run out of entries. Extra emtries are dumped.
+	 */
+	for (i = 0; i < num_rails; i++) {
+		p = rail_info[i];
+		for (j = 0; j < num_mrail_info; j++) {
+			if (!tails[j])
+				mrail_info[j] = p;
+			else
+				tails[j]->next = p;
+			tails[j] = p;
+			p = p->next;
+			tails[j]->next = NULL;
+		}
+		if (p)
+			fi_freeinfo(p);
+	}
```

Comment:
Would this handle the following case?

rail_info:
rail0_info_prov0 -> rail0_info_prov1 -> rail0_info_prov2
rail1_info_prov0 -> rail1_info_prov2

In the above rail1-prov1 combo returned ENODATA for some reason.

Reply:
It would handle, but not in the best way: two mrail info will be returned. the first one maps to rail0_info_prov0 and rail1_info_prov0, the second one maps to rail0_info_prov1 and rail1_info_prov2.  rail0_info_prov2 is dropped.

The old code can encounter the same issue - image rail1_prov0 combo returned ENODATA for some reason...

Reply:
I think depending on application requirement we may want to drop the second mrail_info which maps to rail0_info_prov1 and rail1_info_prov2. If the app is fine with rails being on different providers then we can return it as well.

Yeah, the old code didn't handle many different cases.

Reply:
We can have a follow up PR that use a control knob to determine whether or not to drop such case.

Reply:
👍